## Descriptive and summary statistics

In [20]:
import pandas as pd

In [21]:
df = pd.read_csv('heart_disease.csv')

In [22]:
# # The describe() method displays the descriptive statistics about a DataFrame including the mean, median, min,
# max and quartile values for each numerical column.

df.describe()

age  chest_pain     rest_bp        chol      max_hr     st_depr
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000
mean    54.366337    0.966997  131.623762  246.264026  149.646865    1.039604
std      9.082101    1.032052   17.538143   51.830751   22.905161    1.161075
min     29.000000    0.000000   94.000000  126.000000   71.000000    0.000000
25%     47.500000    0.000000  120.000000  211.000000  133.500000    0.000000
50%     55.000000    1.000000  130.000000  240.000000  153.000000    0.800000
75%     61.000000    2.000000  140.000000  274.500000  166.000000    1.600000
max     77.000000    3.000000  200.000000  564.000000  202.000000    6.200000

In [23]:
# You can also call individual methods on a column (Series object) to get a particular descriptive value
# for that column:

df["age"].min()
# df["chol"].mean()

np.int64(29)

<div class="alert alert-block alert-info">
<b>Tip:</b> To view a summary of non-numerical (categorical) columns, you should set the "include" parameter of the describe method equal to "object". In pandas, categorical variables are of type "object". 
    
    For example:  df.describe(include="object")

The summary will include the top occurring category for each column along with its frequency. 
</div>

In [24]:
df.describe(include="object")

sex target
count      303    303
unique       2      2
top     Female    Yes
freq       207    165

### info()
The info() method displays information about a DataFrame including column data type (dtype) and non-null values.

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   age         303 non-null    int64  
 1   sex         303 non-null    object 
 2   chest_pain  303 non-null    int64  
 3   rest_bp     303 non-null    int64  
 4   chol        303 non-null    int64  
 5   max_hr      303 non-null    int64  
 6   st_depr     303 non-null    float64
 7   target      303 non-null    object 
dtypes: float64(1), int64(5), object(2)
memory usage: 19.1+ KB


<div class="alert alert-block alert-success">
    <b>Note:</b> Using the <b>info()</b> method can be a valuable way to determine if there are any missing values in our dataset. Above, note that info() indicates that the dataset has <b>"303 entries"</b>. Also, note that it indicates that each column has <b>"303 non-null"</b> values. A "null" value indicates a missing value, so the above suggests that there are no missing values in the dataset. Further, info() helpfully provides information about the datatype (Dtype) of each column.</div>

### Displaying the unique values

In [26]:
# The Series object's (a single column) unique() method will return the column's unique values.

df["target"].unique()

array(['Yes', 'No'], dtype=object)

In [27]:
# A Series object (single column) can be passed into Python's built-in set() function 
#   to see the column's unique values.

set(df["target"])

{'No', 'Yes'}

<div class="alert alert-block alert-info">
<b>Uses</b>: We often need to convert categorical variable names into numbers for data analysis and machine learning, so being able to determine the number of unique values within a column can be helpful.  For example, knowing that there are only two categories within a column, as shown above, we would know that we could perhaps convert <b>'Yes'</b> to 1 and <b>'No'</b> to 0, when numbers are required. If we hadn't determined the unique values with the column, perhaps we could have missed that there was a <b>'Maybe'</b> option that also needed to be converted. <div>

## Correlation
The DataFrame's corr() method will display the pairwise correlations among the DataFrame's columns.

In [28]:
df.corr(numeric_only=True)

age  chest_pain   rest_bp      chol    max_hr   st_depr
age         1.000000   -0.068653  0.279351  0.213678 -0.398522  0.210013
chest_pain -0.068653    1.000000  0.047608 -0.076904  0.295762 -0.149230
rest_bp     0.279351    0.047608  1.000000  0.123174 -0.046698  0.193216
chol        0.213678   -0.076904  0.123174  1.000000 -0.009940  0.053952
max_hr     -0.398522    0.295762 -0.046698 -0.009940  1.000000 -0.344187
st_depr     0.210013   -0.149230  0.193216  0.053952 -0.344187  1.000000

<div class="alert alert-block alert-info">
<b>Uses</b>: 
    
    It's important to be able to interpret correlation coefficients. 
    In general, the degree of correlation is as follows:

<b>Perfect</b>: If the value is near ± 1, then as one variable increases, the other variable tends to also increase (if positive) or decrease (if negative).
    
<b>High</b>: If the value lies between ± 0.50 and ± 1, then it is said to be a strong correlation.
    
<b>Moderate</b>: If the value lies between ± 0.30 and ± 0.49, then it is said to be a moderate correlation.
    
<b>Low</b>: When the value lies between ± .29 and 0, then it is said to be a small correlation.
    
<b>No correlation</b>: When the value is zero there is no correlation between the variables.<div>
    
    
   
    Correlation is a very valuable tool when performing data analysis. We will discuss this in more detail  in subsequent modules that cover data analysis. 

In [29]:
# You also can select specific columns of interest and use the corr() method to see the pairwise correlations.

df[["age", "max_hr"]].corr()

age    max_hr
age     1.000000 -0.398522
max_hr -0.398522  1.000000

<div class="alert alert-block alert-success">
<b>Note:</b> Identifying correlations among the columns (independent variables) in a dataset can be very important.  This is known as <b>collinearity</b> and can be detrimental to training a machine learning model.</div>

## Sorting
The DataFrame's sort_values() method takes a "by" parameter to indicate which column the DataFrame should be sorted by:


In [30]:
df.sort_values(by='age')

age     sex  chest_pain  rest_bp  chol  max_hr  st_depr target
72    29  Female           1      130   204     202      0.0    Yes
58    34  Female           3      118   182     174      0.0    Yes
125   34    Male           1      118   210     192      0.7    Yes
65    35    Male           0      138   183     182      1.4    Yes
227   35  Female           0      120   198     130      1.6     No
..   ...     ...         ...      ...   ...     ...      ...    ...
151   71    Male           0      112   149     125      1.6    Yes
60    71    Male           2      110   265     130      0.0    Yes
129   74    Male           1      120   269     121      0.2    Yes
144   76    Male           2      140   197     116      1.1    Yes
238   77  Female           0      125   304     162      0.0     No

[303 rows x 8 columns]

<div class="alert alert-block alert-warning">
<b>Alert:</b> By default, sort_values() will sort the values from <b>smallest to largest</b>.
</div>

<div class="alert alert-block alert-warning">
<b>Alert:</b> Remember, when modifying a DataFrame, pandas typically returns a copy so that the original DataFrame is unchanged. To transfer the modification to the original Dataframe, you can either set the original Dataframe (df) equal to what pandas returns:

    df = df.sort_values(by="age")

Or you can set the method's "inplace" parameter to True so that the Dataframe itself is changed:
    
    df.sort_values(by="age", inplace=True)
    
</div>

In [31]:
# Setting the "ascending" parameter of the sort_values() method to False will sort the values in descending order:

df.sort_values(by="age", ascending=False)

age     sex  chest_pain  rest_bp  chol  max_hr  st_depr target
238   77  Female           0      125   304     162      0.0     No
144   76    Male           2      140   197     116      1.1    Yes
129   74    Male           1      120   269     121      0.2    Yes
25    71    Male           1      160   302     162      0.4    Yes
151   71    Male           0      112   149     125      1.6    Yes
..   ...     ...         ...      ...   ...     ...      ...    ...
65    35    Male           0      138   183     182      1.4    Yes
227   35  Female           0      120   198     130      1.6     No
125   34    Male           1      118   210     192      0.7    Yes
58    34  Female           3      118   182     174      0.0    Yes
72    29  Female           1      130   204     202      0.0    Yes

[303 rows x 8 columns]

### We can also sort by requesting the "nlargest" or "nsmallest" of a given column:

### Get the top n of a feature
Use the nlargest() method to sort by a given column and display n number of rows. Pass in the value of n (the number of rows to display) and use the "column" parameter to indicate which row to sort by:




In [32]:
df[['age', 'max_hr', 'sex']].nlargest(10, columns='max_hr')

age  max_hr     sex
72    29     202  Female
248   54     195  Female
103   42     194  Female
125   34     192    Male
62    52     190  Female
32    44     188  Female
1     37     187  Female
56    48     186  Female
73    51     186  Female
57    45     185  Female

### Get the bottom n of a feature
Use the nsmallest() method to sort by a given column and display n number of rows. Pass in the value of n (the number of rows to display) and use the "column" parameter to indicate which row to sort by:

In [33]:
df[['age', 'sex', 'max_hr']].nsmallest(5, columns='age')

age     sex  max_hr
72    29  Female     202
58    34  Female     174
125   34    Male     192
65    35    Male     182
157   35  Female     174